// @rescript/core/src/Core__Nullable.bs.js
// Generated by ReScript, PLEASE EDIT WITH CARE
'use strict';

var Curry = require("rescript/lib/js/curry.js");
var Caml_option = require("rescript/lib/js/caml_option.js");
var Core__Option = require("./Core__Option.bs.js");

function fromOption(option) {
  if (option !== undefined) {
    return Caml_option.valFromOption(option);
  }
  
}

function equal(a, b, eq) {
  return Core__Option.equal((a == null) ? undefined : Caml_option.some(a), (b == null) ? undefined : Caml_option.some(b), eq);
}

function compare(a, b, cmp) {
  return Core__Option.compare((a == null) ? undefined : Caml_option.some(a), (b == null) ? undefined : Caml_option.some(b), cmp);
}

function getWithDefault(value, $$default) {
  if (value == null) {
    return $$default;
  } else {
    return value;
  }
}

function getExn(value) {
  if (!(value == null)) {
    return value;
  }
  throw {
        RE_EXN_ID: "Invalid_argument",
        _1: "Nullable.getExn: value is null or undefined",
        Error: new Error()
      };
}

function map(value, f) {
  if (value == null) {
    return value;
  } else {
    return Curry._1(f, value);
  }
}

function mapWithDefault(value, $$default, f) {
  if (value == null) {
    return $$default;
  } else {
    return Curry._1(f, value);
  }
}

function flatMap(value, f) {
  if (value == null) {
    return value;
  } else {
    return Curry._1(f, value);
  }
}

exports.equal = equal;
exports.compare = compare;
exports.fromOption = fromOption;
exports.getWithDefault = getWithDefault;
exports.getExn = getExn;
exports.map = map;
exports.mapWithDefault = mapWithDefault;
exports.flatMap = flatMap;
/* No side effect */
